examples/exex/in-memory-state/src/main.rs

```diff
@@ -5,45 +5,144 @@ use reth_exex::{ExExContext, ExExEvent, ExExNotification};
 use reth_node_api::FullNodeComponents;
 use reth_node_ethereum::EthereumNode;
 use reth_tracing::tracing::info;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
 
 /// An ExEx that keeps track of the entire state in memory
-async fn track_state<Node: FullNodeComponents>(mut ctx: ExExContext<Node>) -> eyre::Result<()> {
-    // keeps the entire plain state of the chain in memory
-    let mut state = BundleStateWithReceipts::default();
-
-    while let Some(notification) = ctx.notifications.recv().await {
-        match &notification {
-            ExExNotification::ChainCommitted { new } => {
-                info!(committed_chain = ?new.range(), "Received commit");
-            }
-            ExExNotification::ChainReorged { old, new } => {
-                // revert to block before the reorg
-                state.revert_to(new.first().number - 1);
-                info!(from_chain = ?old.range(), to_chain = ?new.range(), "Received reorg");
-            }
-            ExExNotification::ChainReverted { old } => {
-                state.revert_to(old.first().number - 1);
-                info!(reverted_chain = ?old.range(), "Received revert");
-            }
-        };
+struct InMemoryStateExEx<Node: FullNodeComponents> {
+    /// The context of the ExEx
+    ctx: ExExContext<Node>,
+    /// Entire plain state of the chain
+    state: BundleStateWithReceipts,
+}
+
+impl<Node: FullNodeComponents> InMemoryStateExEx<Node> {
+    /// Create a new instance of the ExEx
+    fn new(ctx: ExExContext<Node>) -> Self {
+        Self { ctx, state: BundleStateWithReceipts::default() }
+    }
+}
 
-        if let Some(committed_chain) = notification.committed_chain() {
-            // extend the state with the new chain
-            state.extend(committed_chain.state().clone());
-            ctx.events.send(ExExEvent::FinishedHeight(committed_chain.tip().number))?;
+impl<Node: FullNodeComponents + Unpin> Future for InMemoryStateExEx<Node> {
+    type Output = eyre::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        while let Some(notification) = ready!(this.ctx.notifications.poll_recv(cx)) {
+            match &notification {
+                ExExNotification::ChainCommitted { new } => {
+                    info!(committed_chain = ?new.range(), "Received commit");
+                }
+                ExExNotification::ChainReorged { old, new } => {
+                    // revert to block before the reorg
+                    this.state.revert_to(new.first().number - 1);
+                    info!(from_chain = ?old.range(), to_chain = ?new.range(), "Received reorg");
+                }
+                ExExNotification::ChainReverted { old } => {
+                    this.state.revert_to(old.first().number - 1);
+                    info!(reverted_chain = ?old.range(), "Received revert");
+                }
+            };
```

Comment:
in this case we should also explain a nice UX for how you'd call async functions inside the future loop?

Reply:
yeah, but I'm still unsure what's the best approach here. So I think we should merge this PR first as it's about the testing, and then mention what you said in the book.